src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -0,0 +1,143 @@
+import com.atlassian.bitbucket.jenkins.internal.scm.BitbucketSCMSource;
+import hudson.Extension;
+import hudson.model.Item;
+import hudson.plugins.git.BranchSpec;
+import hudson.scm.SCM;
+import hudson.scm.SCMDescriptor;
+import hudson.util.FormValidation.Kind;
+import jenkins.plugins.git.GitBranchSCMHead;
+import jenkins.scm.api.*;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+import static org.eclipse.jgit.lib.Constants.*;
+
+public class BitbucketSCMFileSystem extends SCMFileSystem {
+
+    private final BitbucketFilePathClient client;
+    private final String ref;
+
+    protected BitbucketSCMFileSystem(BitbucketFilePathClient client, @Nullable SCMRevision scmRevision,
+                                     @Nullable String ref) {
+        super(scmRevision);
+        this.client = client;
+        this.ref = ref;
+    }
+
+    @Override
+    public SCMFile getRoot() {
+        return new BitbucketSCMFile(client, ref);
+    }
+
+    // We do not provide this information in the REST response, so this is undefined.
+    @Override
+    public long lastModified() throws IOException, InterruptedException {
+        return 0;
+    }
+
+    @Extension
+    public static class BuilderImpl extends SCMFileSystem.Builder {
+
+        @Inject
+        BitbucketClientFactoryProvider clientFactoryProvider;
+        @Inject
+        JenkinsToBitbucketCredentials jenkinsToBitbucketCredentials;
+        @Inject
+        BitbucketPluginConfiguration pluginConfiguration;
+
+        @Override
+        public SCMFileSystem build(Item item, SCM scm,
+                                   @CheckForNull SCMRevision scmRevision) throws IOException, InterruptedException {
+            if (!(scm instanceof BitbucketSCM)) {
+                return null;
+            }
+
+            BitbucketSCM bitbucketSCM = (BitbucketSCM) scm;
+            Optional<BitbucketServerConfiguration> maybeServerConfiguration =
+                    pluginConfiguration.getServerById(bitbucketSCM.getServerId());
+            if (!maybeServerConfiguration.isPresent() || maybeServerConfiguration.get().validate().kind == Kind.ERROR) {
+                return null;
+            }
+
+            BitbucketSCMRepository repository = bitbucketSCM.getBitbucketSCMRepository();
+
+            BitbucketFilePathClient filePathClient =
+                    clientFactoryProvider.getClient(maybeServerConfiguration.get().getBaseUrl(),
+                            jenkinsToBitbucketCredentials.toBitbucketCredentials(repository.getCredentialsId()))
+                            .getProjectClient(repository.getProjectKey())
+                            .getRepositoryClient(repository.getRepositorySlug())
+                            .getFilePathClient();
+
+            return new BitbucketSCMFileSystem(filePathClient, null, bitbucketSCM.getBranches().get(0).toString());
```

Comment:
We use the branch selector as the ref and _only_ permit refs/heads and refs/tags because pipelines have no awareness of the branch they're on (if, for example, your selector is **, and you hit "build now" on your first pipeline build, Jenkins can't tell you where to find a Jenkinsfile). The SCMRevision for pipeline jobs is _always_ null.

Reply:
It'd be nice to add this comment in the code too